pkg/client/opensergo_client_observer.go

```diff
@@ -0,0 +1,252 @@
+	return &OpensergoClientObserver{
+		subscribeResponseChan: make(chan transportPb.SubscribeResponse, 1024),
+		opensergoClientStatus: opensergoClientStatus,
+	}
+}
+
+func (opensergoClientObserver *OpensergoClientObserver) Start() {
+
+	if opensergoClientObserver.receiveGoroutineOpened.Get() == 0 {
+		// keepalive OpensergoClient
+		logging.Info("open observeReceive() goroutine to observe msg from opensergo-control-plane to subscribeResponseChannel.")
+		go opensergoClientObserver.observeReceive()
+	}
+	if opensergoClientObserver.handleGoroutineOpened.Get() == 0 {
+		// keepalive OpensergoClient
+		logging.Info("open doHandle() goroutine to handle msg in subscribeResponseChannel.")
+		go opensergoClientObserver.doHandle()
+	}
+}
+
+// observeReceive receive message from opensergo-control-plane.
+//
+// Contains a for-loop without any interrupt-exit.
+// Push the subscribeResponse into a channel in OpensergoClientObserver
+func (opensergoClientObserver *OpensergoClientObserver) observeReceive() {
+	opensergoClientObserver.receiveGoroutineOpened = groupcache.AtomicInt(1)
+	var errorLocal error
+	if opensergoClientObserver.subscribeConfigClient == nil {
+		errorLocal = errors.New("can not doReceive(), grpc client subscribeConfigClient is nil")
+	}
+	subscribeResponse, err := (*opensergoClientObserver.subscribeConfigClient).Recv()
+	if err != nil {
+		errorLocal = err
+	}
+	if errorLocal != nil {
+		logging.Error(errorLocal, "error when receive config-data.")
+		opensergoClientObserver.subscribeConfigClient = nil
+		opensergoClientObserver.opensergoClientStatus.Store(interrupted)
+		opensergoClientObserver.receiveGoroutineOpened = groupcache.AtomicInt(0)
+		logging.Warn("end observeReceive()")
+		return
+	}
+
+	opensergoClientObserver.subscribeResponseChan <- *subscribeResponse
+
+	opensergoClientObserver.observeReceive()
+}
+
+// doHandle handle the received messages from opensergo-control-plane.
+//
+// Contains a select to block this action
+func (opensergoClientObserver *OpensergoClientObserver) doHandle() {
+	opensergoClientObserver.handleGoroutineOpened = groupcache.AtomicInt(1)
+
+	subscribeResponse := <-opensergoClientObserver.subscribeResponseChan
+	handleReceive(subscribeResponse, *opensergoClientObserver.subscribeConfigClient)
+
+	opensergoClientObserver.doHandle()
+}
+
+// handleReceive handle the received response from opensergo-control-plane.
+func handleReceive(subscribeResponse transportPb.SubscribeResponse, subscribeConfigClient transportPb.OpenSergoUniversalTransportService_SubscribeConfigClient) {
+
+	ack := subscribeResponse.GetAck()
+	// response of client-send
+	if ack != "" {
+		code := subscribeResponse.GetStatus().GetCode()
+		if transport.CODE_SUCCESS == code {
+			return
+		}
+
+		if code >= transport.CODE_ERROR_UNKNOWN && code < transport.CODE_ERROR_UPPER_BOUND {
+			// TODO log info code out of bound
+			return
+		}
+	}
+
+	// handle message of server-push
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
recover recommends printing stack information and error causes

Reply:
fixed! thans for you review ~